src/kakarot/precompiles/kakarot_precompiles.cairo

```diff
@@ -87,6 +90,50 @@ namespace KakarotPrecompiles {
         }
 
         let (revert_reason_len, revert_reason) = Errors.invalidCairoSelector();
-        return (revert_reason_len, revert_reason, CAIRO_PRECOMPILE_GAS, 1);
+        return (revert_reason_len, revert_reason, CAIRO_PRECOMPILE_GAS, Errors.EXCEPTIONAL_HALT);
+    }
+
+    func cairo_message{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(evm_address: felt, input_len: felt, input: felt*) -> (
+        output_len: felt, output: felt*, gas_used: felt, reverted: felt
+    ) {
+        alloc_locals;
+
+        // Input must be at least 3*32 bytes long.
+        let is_input_invalid = is_le(input_len, 95);
+        if (is_input_invalid != 0) {
+            let (revert_reason_len, revert_reason) = Errors.outOfBoundsRead();
+            return (revert_reason_len, revert_reason, CAIRO_MESSAGE_GAS, Errors.EXCEPTIONAL_HALT);
+        }
+
+        // Input is formatted as:
+        // [to_address: address][data_offset: uint256][data_len: uint256][data: bytes[]]
+
+        // Load target EVM address
+        let target_address = Helpers.bytes32_to_felt(input);
+
+        let data_offset_ptr = input + 32;
+        let data_len_offset = Helpers.bytes32_to_felt(data_offset_ptr);
+        let data_len_ptr = input + data_len_offset;
```

Comment:
we don't really need this as we assume that the payload as the given form in comment, and assert for len >= 3*32, so I'd just go directly to input + 2*32